### PR TITLE
Implement read-write splitting performance tests

### PR DESCRIPTION
**Pull request by joyc-bq**

### CODE REVIEW

File: wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    }
+  }
+
+  protected static Properties initNoPluginPropsWithTimeouts() {
+    final Properties props = new Properties();
+    props.setProperty(PGProperty.USER.getName(), AURORA_POSTGRES_USERNAME);
+    props.setProperty(PGProperty.PASSWORD.getName(), AURORA_POSTGRES_PASSWORD);
+    props.setProperty(PGProperty.TCP_KEEP_ALIVE.getName(), Boolean.FALSE.toString());
+    props.setProperty(PGProperty.CONNECT_TIMEOUT.getName(), "30");
+    props.setProperty(PGProperty.SOCKET_TIMEOUT.getName(), "30");
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java:117]: Do these timeouts need to be 30 seconds? Tests might run faster if you lower these values.

File: wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    }
+    final long setReadOnlyTrueMin = elapsedSetReadOnlyTrueTimes.stream().min(Long::compare).orElse(0L);
+    final long setReadOnlyTrueMax = elapsedSetReadOnlyTrueTimes.stream().max(Long::compare).orElse(0L);
+    final long setReadOnlyTrueAvg =
+        (long) elapsedSetReadOnlyTrueTimes.stream().mapToLong(a -> a).summaryStatistics().getAverage();
+
+    final long setReadOnlyFalseMin = elapsedSetReadOnlyFalseTimes.stream().min(Long::compare).orElse(0L);
+    final long setReadOnlyFalseMax = elapsedSetReadOnlyFalseTimes.stream().max(Long::compare).orElse(0L);
+    final long setReadOnlyFalseAvg =
+        (long) elapsedSetReadOnlyFalseTimes.stream().mapToLong(a -> a).summaryStatistics().getAverage();
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java:160]: Will you get the same result by storing the summaryStatistics first then just calling getMax(), getMin() and getAverage()?
```java
    final LongSummaryStatistics stats = elapsedSetReadOnlyFalseTimes.stream().mapToLong(a -> a).summaryStatistics();
    stats.getMax();
    stats.getMin();
    stats.getAverage();
```

File: wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    public long setReadOnlyTrueMin;
+    public long setReadOnlyTrueMax;
+    public long setReadOnlyTrueAvg;
+
+    public long setReadOnlyFalseMin;
+    public long setReadOnlyFalseMax;
+    public long setReadOnlyFalseAvg;
+  }
+
+  private abstract class PerfStatBase {
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java:284]: Result and PerfStatBase may be static classes.

File: wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+    final long minAverageExecuteStatementTime =
+        elapsedReaderSwitchExecuteStatementsTimes.stream().min(Long::compare).orElse(0L);
+    final long maxAverageReaderSwitchExecuteStatementTime =
+        elapsedReaderSwitchExecuteStatementsTimes.stream().max(Long::compare).orElse(0L);
+    final long avgAverageExecuteStatementTime =
+        (long) elapsedReaderSwitchExecuteStatementsTimes.stream().mapToLong(a -> a).summaryStatistics().getAverage();
+
+    data.minAverageExecuteStatementTime =
+        round((minAverageExecuteStatementTime / (double) TimeUnit.MILLISECONDS.toNanos(1)), 3);
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java:266]: You can store `(double) TimeUnit.MILLISECONDS.toNanos(1)` to a constant to reduce code duplication.

File: wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    final Properties props = initNoPluginPropsWithTimeouts();
+    props.setProperty(PropertyDefinition.PLUGINS.name, "auroraHostList,readWriteSplitting");
+    props.setProperty(ReadWriteSplittingPlugin.LOAD_BALANCE_READ_ONLY_TRAFFIC.name, "true");
+    return props;
+  }
+
+  Result getSetReadOnlyResults(final Properties props)
+      throws SQLException {
+    Long setReadOnlyTrueStartTime;
+    final List<Long> elapsedSetReadOnlyTrueTimes = new ArrayList<>(REPEAT_TIMES);
```
[the remark this answers is not in the document]
Comment (joyc-bq) [reply wrapper/src/test/java/integration/container/aurora/postgres/AuroraPostgresReadWriteSplittingPerformanceTest.java:135]: "Type argument cannot be of primitive type"

**Review by congoamz — APPROVED** [review]

**Pull request merged**